### PR TITLE
Queues: Move Notification to use a Queue

### PR DESCRIPTION
**Pull request by zeripath**
Move the notifier system to use a Queue.

* First step is to move the ui service to use a Queue
* Next in preparation to use a Queue we ensure that we can json marshal things passed in the notification
* Autogeneration is used to create a notifier that wraps all the notification calls to a queue push and expand them back out
* A special notification helper is provided for tests allowing these to specifically check notification events

Closes #8874

### CODE REVIEW

**Comment by zeripath** [comment]
This one is potentially quite controversial ... Basically every request can cause a notification so this one might require a large amount of workers if we don't want it to get overloaded.

**Comment by guillep2k** [comment]
> 
> 
> This one is potentially quite controversial ... Basically every request can cause a notification so this one might require a large amount of workers if we don't want it to get overloaded.

He he... 😁

Well, I'm in favor of longer queues rather than many workers. If we've got to send 1,000 mails or insert as many UI notifications, I'd rather do it in one or two workers than in 1,000. Heck! Our default db pool size is _unlimited_, which is not very friendly.

And since all workers would be potentially competing for the same resources, I think less workers would eventually be much faster if the number of items is sufficiently large. Maybe with this PR it will be easier to test, if I generate a large number of users. 🤔

2¢

**Comment by zeripath** [comment]
So there's still a race in the pull merge test. I'll have to do a bit more work on that.

I guess it's just enforce that what we're seeing is actually a merge or not for this test.

**Comment by zeripath** [comment]
@guillep2k I agree, we should grow worker pools where we know that they can cope with the concurrency and restrict the others with a large enough channel (for channel based queues.)

File: models/issue.go [hunk 1]
```diff
@@ -47,7 +47,7 @@ type Issue struct {
 	IsClosed         bool         `xorm:"INDEX"`
 	IsRead           bool         `xorm:"-"`
 	IsPull           bool         `xorm:"INDEX"` // Indicates whether is a pull request or not.
-	PullRequest      *PullRequest `xorm:"-"`
+	PullRequest      *PullRequest `xorm:"-" json:"-"`
```
Comment (guillep2k) [inline models/issue.go:50]: I'm wondering whether attachments, comments, reactions, assignees, etc. should be skipped too. Because there are reference fields in those structs as well (isn't there an `Issue` property in `Comment`?). Sounds like recursions can cause some headaches. On the other hand, the serialized data is a snapshot of the object state that protects the notifier from other processes changing the data, so it makes some sense to keep it as is. Decisions, decisions... 🤔

File: integrations/notification_helper_test.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	n.lock.Lock()
+	found := -1
+	for i, callbackPtr := range n.callbacks[functionName] {
+		if callbackPtr == callback {
+			found = i
+			break
+		}
+	}
+	if found > -1 {
+		n.callbacks[functionName] = append(n.callbacks[functionName][0:found], n.callbacks[functionName][found+1:]...)
```
Comment (guillep2k) [inline integrations/notification_helper_test.go]: Since you will be breaking from the loop, I think it's safe to omit the `found` variable and move this inside `if callbackPtr == callback`.

File: integrations/pull_merge_test.go [hunk 3]
```diff
@@ -61,31 +61,75 @@ func testPullCleanUp(t *testing.T, session *TestSession, user, repo, pullnum str
 
 func TestPullMerge(t *testing.T) {
 	onGiteaRun(t, func(t *testing.T, giteaURL *url.URL) {
-		hookTasks, err := models.HookTasks(1, 1) //Retrieve previous hook number
-		assert.NoError(t, err)
-		hookTasksLenBefore := len(hookTasks)
+		createPullNotified, deferableCreate := notifierListener.RegisterChannel("NotifyNewPullRequest", 0, &models.PullRequest{})
```
Comment (guillep2k) [inline integrations/pull_merge_test.go]:
```suggestion
		createPullNotified, unregisterCreate := notifierListener.RegisterChannel("NotifyNewPullRequest", 0, &models.PullRequest{})
```

I think that `deferableCreate` is confusing. 😕

File: integrations/pull_merge_test.go [hunk 4]
```diff
@@ -61,31 +61,75 @@ func testPullCleanUp(t *testing.T, session *TestSession, user, repo, pullnum str
...
+		pr.LoadHeadRepo()
+		pr.BaseRepo.MustOwner()
+		pr.HeadRepo.MustOwner()
+
+		assert.EqualValues(t, "user1", pr.HeadRepo.Owner.Name)
+		assert.EqualValues(t, "repo1", pr.HeadRepo.Name)
+		assert.EqualValues(t, "user2", pr.BaseRepo.Owner.Name)
+		assert.EqualValues(t, "repo1", pr.BaseRepo.Name)
+
+		time.Sleep(100 * time.Millisecond)
```
Comment (guillep2k) [inline integrations/pull_merge_test.go]: Why is this `Sleep` required?

File: integrations/pull_merge_test.go [hunk 5]
```diff
@@ -96,20 +140,18 @@ func TestPullRebase(t *testing.T) {
...
+		case <-mergePullNotified:
+		case <-time.After(500 * time.Millisecond):
+			assert.Fail(t, "Took too long to notify!")
+		}
 	})
 }
 
 func TestPullRebaseMerge(t *testing.T) {
 	onGiteaRun(t, func(t *testing.T, giteaURL *url.URL) {
-		defer prepareTestEnv(t)()
```
Comment (guillep2k) [inline integrations/pull_merge_test.go]: Why are these no longer required?

**Review by guillep2k — COMMENTED** [review]
Oh, man, you've been having fun! 😁

File: integrations/pull_merge_test.go (hunk 5 above)
Comment (zeripath) [reply integrations/pull_merge_test.go]: They never were required. onGiteaRun has had a prepareTestEnv already.

File: modules/notification/base/notifier.go [hunk 6]
```diff
@@ -9,6 +9,8 @@ import (
 	"code.gitea.io/gitea/modules/repository"
 )
 
+//go:generate go run -mod=vendor main.go
```
[the remark this answers is not in the document]
Comment (zeripath) [reply modules/notification/base/notifier.go:12]: Yeah or just make build as we run go generate in build now. 

notifier.go needs a comment to this effect though.

File: models/issue.go (hunk 1 above)
Comment (zeripath) [reply models/issue.go:50]: yup. I decided to bridge the problem as and when we see it. The trouble is under the default testing case you might not see the issue as we don't necessarily try to jsonify these things. (Hmm maybe I should add a jsontestqueue that will always json.Marshal stuff.)

As we pass into notifier pull requests and issues which definitely have a cyclic reference and therefore MarshalJSON doesn't work - we need to break the cycle. Every PR has an issue. Not all issues have PRs - so breaking the cycle at issue seemed best.

File: modules/notification/base/main.go [hunk 7]
```diff
@@ -0,0 +1,246 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (zeripath) [reply modules/notification/base/main.go]: Done

File: modules/notification/base/main.go [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+}
+
+func main() {
+	fset := token.NewFileSet() // positions are relative to fset
+	f, err := parser.ParseFile(fset, "notifier.go", nil, 0)
+	if err != nil {
+		panic(err)
+	}
+	funcs := make([]funcDef, 0)
+	//currentFunc := funcDef{}
```
[the remark this answers is not in the document]
Comment (zeripath) [reply modules/notification/base/main.go]: Done

File: modules/notification/base/notifier.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zeripath) [reply modules/notification/base/notifier.go:12]: Added a few comments

File: modules/notification/base/main.go [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+					}
+				}
+			}
+			funcs[i] = def
+		}
+
+		return true
+	})
+
+	buf := bytes.Buffer{}
```
[the remark this answers is not in the document]
Comment (zeripath) [reply modules/notification/base/main.go:135]: Done.

File: integrations/notification_helper_test.go (hunk 2 above)
Comment (zeripath) [reply integrations/notification_helper_test.go]: Done

File: integrations/pull_merge_test.go (hunk 3 above)
Comment (zeripath) [reply integrations/pull_merge_test.go]: Done

File: integrations/pull_merge_test.go (hunk 4 above)
Comment (zeripath) [reply integrations/pull_merge_test.go]: I've changed this whole section of the test to be a bit clearer.

File: integrations/notification_helper_test.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (zeripath) [reply integrations/notification_helper_test.go]: Done

File: modules/notification/base/notifier.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (lunny) [reply modules/notification/base/notifier.go:12]: Why we need to generate that two files? Any advantage the new method than before?

**Comment by lunny** [comment]
For test with notification, I think we just need a mockNotification and register only this notification. That will make the code simpler.

File: models/issue.go (hunk 1 above)
Comment (guillep2k) [reply models/issue.go:50]: I'm not particularly worried about cyclic references; those should stand out soon enough. But I think the overhead of serializing the whole bunch might be excessive. What I'm thinking is to use the opposite strategy: pass on only what is not derived, unless there's a reason to.

On the other hand, passing data which is tightly-coupled with the running Gitea version poses a problem for upgrading if any queues contain data during the process (e.g. file-system backed queues retained across shutdowns). One possible solution to this problem is to prevent upgrading if any queues contain data, but that may require a migration step being aware of the queue configuration and it would be very difficult to implement. May be we could drop a record on the database when shutting down if all queues are empty and the shutdown was orderly? A migration step could refuse to continue if that record is missing.

File: modules/notification/base/notifier.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zeripath) [reply modules/notification/base/notifier.go:12]: 1. If you change notifier.go they will be both be kept up to date.
2. It would be beyond tiresome to write the notifier queue by hand and incredibly difficult to keep up to date.

File: models/issue.go (hunk 1 above)
Comment (zeripath) [reply models/issue.go:50]: The trouble with not passing "derived" data is that the notification you may get may be different from what was done.

You have to pass and parse all the data that makes the notification because you can't be certain what's going to change in future.

If you migrate and you can't deserialize mostly the notification will be dropped or default values added. 

As you say we need a flush mechanism. That may mean opening a non listening Gitea to run through its queues.

**Review by guillep2k — APPROVED** [review]

**Comment by zeripath** [comment]
I've forcibly rebased on #10001 as this should fix a number of issues with the tests.

File: routers/api/v1/repo/issue_comment.go [hunk 11]
```diff
@@ -413,6 +413,8 @@ func editIssueComment(ctx *context.APIContext, form api.EditIssueCommentOption)
 		ctx.Error(http.StatusInternalServerError, "UpdateComment", err)
 		return
 	}
+	_ = comment.LoadAssigneeUser()
```
Comment (lunny) [inline routers/api/v1/repo/issue_comment.go]: This change is unrelated.

File: modules/notification/ui/ui.go [hunk 12]
```diff
@@ -29,19 +31,24 @@ var (
 
 // NewNotifier create a new notificationService notifier
 func NewNotifier() base.Notifier {
```
Comment (lunny) [inline modules/notification/ui/ui.go]: I think this change could be another smaller PR which should be easier to review.

File: routers/api/v1/repo/issue_comment.go (hunk 11 above)
Comment (zeripath) [reply routers/api/v1/repo/issue_comment.go]: done

File: modules/notification/ui/ui.go (hunk 12 above)
Comment (zeripath) [reply modules/notification/ui/ui.go]: the single commit ebdebd77921ae49102fd613f2a00b6e514a35e62 has been extracted as #10281

**Comment by 6543** [comment]
can you pleace resolve confilcts :D

**Comment by lafriks** [comment]
@zeripath ping

**Comment by lunny** [comment]
We have to be careful most notifier function params are pointers that maybe changed in the notifiers' go routines or the original go routine.

**Review by lafriks — APPROVED** [review]

**Comment by zeripath** [comment]
This should not go in to 1.12 at this stage.

**Comment by 6543** [comment]
pleace resolve conflict
